File: Count Number of Maximum Bitwise-OR Subsets.cpp
class Solution {
public:

    int solve (int idx, int curror, vector<int>& nums, int& maxor, vector<vector<int>>& t) {
        if (idx == nums.size()) return (curror == maxor) ? 1 : 0;
        if (t[curror][idx] != -1) return t[curror][idx];
        int pick = solve(idx + 1, curror | nums[idx], nums, maxor, t);
        int notpick = solve(idx + 1, curror, nums, maxor, t);
        return t[curror][idx] = pick + notpick;
    }

    int countMaxOrSubsets(vector<int>& nums) {
        int maxor = 0;
        for (const int& num: nums) maxor |= num;
        vector<vector<int>> t(maxor + 1, vector<int>(nums.size() + 1, -1));
        return solve(0, 0, nums, maxor, t);
    }
};

File: Count Number of Maximum Bitwise-OR Subsets/Count Number of Maximum Bitwise-OR Subsets.java
// Solution for Count Number of Maximum Bitwise-OR Subsets in JAVA

class Solution {
    public int countMaxOrSubsets(int[] nums) {
        int n = nums.length;
        int maxOr = 0;
        for (int num : nums) {
            maxOr |= num;
        }

        Integer[][] memo = new Integer[n + 1][maxOr + 1];

        return solve(0, 0, maxOr, nums, memo);
    }

    private int solve(int idx, int currOr, int maxOr, int[] nums, Integer[][] memo) {
        if (idx == nums.length) {
            return currOr == maxOr ? 1 : 0;
        }
        if (memo[idx][currOr] != null) {
            return memo[idx][currOr];
        }

        int take = solve(idx + 1, currOr | nums[idx], maxOr, nums, memo);
        int notTake = solve(idx + 1, currOr, maxOr, nums, memo);

        memo[idx][currOr] = take + notTake;
        return memo[idx][currOr];
    }
}
